src/openassetio-core/src/utils/formatter.cpp

```diff
@@ -0,0 +1,198 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2023 The Foundry Visionmongers Ltd
+
+#include "formatter.hpp"
+
+#include <algorithm>
+#include <sstream>
+#include <utility>
+
+#include <openassetio/Context.hpp>
+#include <openassetio/EntityReference.hpp>
+#include <openassetio/errors/BatchElementError.hpp>
+#include <openassetio/hostApi/Manager.hpp>
+#include <openassetio/managerApi/HostSession.hpp>
+#include <openassetio/managerApi/ManagerInterface.hpp>
+#include "../errors/exceptionMessages.hpp"
+
+auto fmt::formatter<openassetio::EntityReference>::format(
+    const openassetio::EntityReference& entityRef, format_context& ctx) const
+    -> decltype(ctx.out()) {
+  return fmt::formatter<string_view>::format(fmt::format("{}", entityRef.toString()), ctx);
+}
+
+auto fmt::formatter<openassetio::EntityReferences>::format(
+    const openassetio::EntityReferences& entityRefs, format_context& ctx) const
+    -> decltype(ctx.out()) {
+  std::vector<std::string> quotedRefs;
+  quotedRefs.reserve(entityRefs.size());
+  std::transform(
+      entityRefs.cbegin(), entityRefs.cend(), std::back_inserter(quotedRefs),
+      [](const openassetio::EntityReference& ref) { return fmt::format("'{}'", ref.toString()); });
+
+  return fmt::formatter<string_view>::format(fmt::format("[{}]", fmt::join(quotedRefs, ", ")),
```

Comment:
I'm looking at that `join` call and don't understand it at all. Does it even take 3 arguments, I thought you could only put in a seperator?

Reply:
There are just two arguments, the final "argument", i.e. comma,  is in quotes. I'll try to break it down

```c++
if (entityRefs.empty()) {
  return "[]";
}

std::string result;
result += "['";
result += fmt::join(entityRefs, "', '");
result += "']"
```

I think that works?

Reply:
Oh _I see_, yeah not a bad idea.

This may be quicker, i'd have to profile, although I suspect it's going to be equivalent at best. Temporary vector and Temporary string are similar in performance cost imo, but the real killer is the need to convert the output of `join` to a string.

:shrug: If  `fmt::format` itself is quite slow, you could make this argument, but I think without any metrics either implementation has a shot at being faster, so just leaving this as it is.

Reply:
https://github.com/OpenAssetIO/OpenAssetIO/pull/1329/commits/8a6fff22ba018148b36f05e2a06f78b4307f715d